OpenRA.Mods.Common/Traits/World/AutoSave.cs

```diff
@@ -0,0 +1,88 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[TraitLocation(SystemActors.World)]
+	[Desc("Add this trait to the world actor to enable auto-save.")]
+	public class AutoSaveInfo : TraitInfo
+	{
+		public override object Create(ActorInitializer init) { return new AutoSave(init.Self, this); }
+	}
+
+	public class AutoSave : ITick
+	{
+		const string AutoSavePattern = "auto-save@";
+		const string SaveFileExtension = ".orasav";
+		readonly DirectoryInfo autoSaveDirectoryInfo;
+		readonly string saveFolderPath;
+		uint ticksUntilAutoSave;
+
+		public AutoSave(Actor self, AutoSaveInfo info)
+		{
+			ticksUntilAutoSave = (uint)GetTicksBetweenAutosaves(self);
+
+			var mod = Game.ModData.Manifest;
+
+			saveFolderPath = Path.Combine(Platform.SupportDir, "Saves", mod.Id, mod.Metadata.Version);
+
+			autoSaveDirectoryInfo = new DirectoryInfo(saveFolderPath);
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			var autoSaveValue = Game.Settings.Gameplay.AutoSaveInterval;
+			var autoSaveFileLimit = Game.Settings.Gameplay.AutoSaveFileNumber;
+
+			autoSaveFileLimit = autoSaveFileLimit < 3 ? 3 : autoSaveFileLimit;
+
+			if (self.World.Paused || autoSaveValue == 0 || autoSaveFileLimit == 0) return;
+
+			if (ticksUntilAutoSave == 0)
+				ticksUntilAutoSave = (uint)GetTicksBetweenAutosaves(self);
+
+			if (--ticksUntilAutoSave > 0)
+				return;
+
+			if (GetAutoSaveFiles().Count() > autoSaveFileLimit)
```

Comment:
please stop refusing to change this function, you are requesting all the files from the file system several times... https://github.com/OpenRA/OpenRA/pull/21693#discussion_r1912010429

 Also before executing you need to check wether the directory even exists. Currently it doesn't the game crashes. We should skip deleting if it doesn't exist, the saving logic will create it eventually

```
if (Directory.Exists(saveFolderPath))
```